Enforce doubly-linked chained block

qwe661234 opened the pull request:
Previously, the chained block was a linear structure where previous block only pointed to the next block. Now, we have introduced additional information, allowing the next block to also reference the previous block. This modification enhance the deletion of replaced block.

Close: #329

qwe661234 committed:
Remove predicted block

Because we already have block chaining, there is no need for predicted
block.

jserv commented on src/utils.h:
```diff
@@ -56,3 +56,65 @@ static inline uintptr_t align_up(uintptr_t sz, size_t alignment)
+    struct list_head *prev, *next;
+};
+
+inline void INIT_LIST_HEAD(struct list_head *head)
```
Use `static inline` rather than `inline`.

jserv commented on src/utils.h:
```diff
@@ -56,3 +56,65 @@ static inline uintptr_t align_up(uintptr_t sz, size_t alignment)
+    head->prev = head;
+}
+
+inline int list_empty(const struct list_head *head)
```
Specify `bool` for the return type.

jserv commented on src/emulate.c:
```diff
@@ -1016,16 +1052,19 @@ void rv_step(riscv_t *rv, int32_t cycles)
+        /* lookup the next block in block map or translate a new block,
+         * and move onto the next block.
+         */
+        block = block_find_or_translate(rv);
```
Declare `block` here.

jserv commented on src/emulate.c:
```diff
@@ -1016,16 +1052,19 @@ void rv_step(riscv_t *rv, int32_t cycles)
-             * and move onto the next block.
-             */
-            block = block_find_or_translate(rv);
+        if (prev) {
```
Rewrite as following:
```c
        if (prev && prev->pc_start != last_pc) {
            /* update previous block */
#if !RV32_HAS(JIT)
            prev = block_find(&rv->block_map, last_pc);
#else
            prev = cache_get(rv->block_cache, last_pc);
#endif
        }
```

jserv commented on src/emulate.c:
```diff
@@ -1034,28 +1073,41 @@ void rv_step(riscv_t *rv, int32_t cycles)
                     last_ir->branch_taken = block->ir_head;
+#if RV32_HAS(JIT)
+                    chaining_entry_t *new_entry =
+                        malloc(sizeof(chaining_entry_t));
```
Shall we utilize memory pool when possible?

jserv commented on src/emulate.c:
```diff
@@ -979,6 +979,49 @@ static block_t *block_find_or_translate(riscv_t *rv)
+                    target->branch_untaken = NULL;
+                free(entry);
+            }
+            /* free delete block */
```
Fix the grammar mistake.

jserv commented on src/utils.h:
```diff
@@ -56,3 +56,65 @@ static inline uintptr_t align_up(uintptr_t sz, size_t alignment)
+
+inline void INIT_LIST_HEAD(struct list_head *head)
+{
+    head->next = head;
```
Rewrite as `head->next = head->prev = head;`

jserv reviewed (changes requested):
Explicitly mention `Close #329` in the body of git commit messages.

jserv commented on src/riscv_private.h:
```diff
@@ -68,9 +68,17 @@ typedef struct block {
+typedef struct chaining_entry {
+    block_t *block;
+    struct list_head list;
+} chaining_entry_t;
```
Simply name it `chain_entry_t`.

jserv commented on src/riscv_private.h:
```diff
@@ -68,9 +68,17 @@ typedef struct block {
 } block_t;
 
+#if RV32_HAS(JIT)
+typedef struct chaining_entry {
```
Use anonymous struct.

jserv commented on src/riscv_private.h:
```diff
@@ -117,6 +125,7 @@ struct riscv_internal {
     block_map_t block_map; /**< basic block map */
 #else
     struct cache *block_cache;
+    struct mpool *chaining_entry_mp;
```
Ditto. Use shorter form. i.e., `chain_entry_pool`.

jserv reviewed (changes requested):
Be aware of the terminology. It is referred as "doubly-linked chained block" instead of "doubly-link chained block."

jserv commented on src/emulate.c:
```diff
@@ -979,6 +979,49 @@ static block_t *block_find_or_translate(riscv_t *rv)
+            if (prev == delete_target)
+                prev = NULL;
+            chain_entry_t *entry, *safe;
+            /* correctly remove delete block from its chained block */
```
should be "deleted block"

jserv commented on src/emulate.c:
```diff
@@ -979,6 +979,49 @@ static block_t *block_find_or_translate(riscv_t *rv)
+                }
+                assert(flag);
+            }
+            /* correctly remove delete block from the block chained to it */
```
Ditto. "deleted block"

jserv commented on src/jit.c:
```diff
@@ -1496,14 +1496,14 @@ static void translate_chained_block(struct jit_state *state,
-        if (block1 && block1->translatable)
+    if (ir->branch_taken && !set_has(set, ir->branch_taken->pc)) {
+        block_t *block1 = cache_get(rv->block_cache, ir->branch_taken->pc);
+        if (block1->translatable)
```
Shall we check `block1` before dereferencing?

qwe661234 replied:
The block1 must be existed because we correctly handle deleted block.

qwe661234 committed:
Enforce doubly-linked chained block

Previously, the chained block was a linear structure where previous
block only pointed to the next block. Now, we have introduced additional
information, allowing the next block to also reference the previous
block. This modification enhance the deletion of replaced block.

Close: #329

visitorckw commented:
LGTM, the patches indeed addressed the issue, thanks!
